src/csharp/Grpc.Core/Internal/BinaryStringDictionary.cs

```diff
@@ -0,0 +1,130 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+using Grpc.Core.Utils;
+
+namespace Grpc.Core.Internal
+{
+    /// <summary>
+    /// Dictionary of string instances that can be looked up by their binary
+    /// representation (ASCII encoded).
+    /// </summary>
+    internal class BinaryStringDictionary
+    {
+        static readonly Encoding EncodingACII = System.Text.Encoding.ASCII;
+        Dictionary<BinaryKey, string> dict = new Dictionary<BinaryKey, string>();
+
+        public BinaryStringDictionary()
+        {
+        }
+
+        public void Add(string str)
+        {
+            dict.Add(new BinaryKey(EncodingACII.GetBytes(str)), str);
+        }
+
+        /// <summary>
+        /// Looks up the string based on its ASCII-encoded binary form.
+        /// </summary>
+        public string Lookup(IntPtr nativeData, int nativeDataLen)
+        {
+            var key = new BinaryKey(nativeData, nativeDataLen);
+            dict.TryGetValue(key, out string str);
+            return str;
+        }
+        
+        private struct BinaryKey : IEquatable<BinaryKey>
+        {
+            private byte[] data;
+            private IntPtr nativeData;
+            private int nativeDataLen;
+
+            public BinaryKey(byte[] data)
+            {
+                this.data = GrpcPreconditions.CheckNotNull(data);
+                this.nativeData = IntPtr.Zero;
+                this.nativeDataLen = 0;
+            }
+
+            public BinaryKey(IntPtr nativeData, int nativeDataLen)
+            {
+                GrpcPreconditions.CheckArgument(nativeData != IntPtr.Zero);
+                GrpcPreconditions.CheckArgument(nativeDataLen >= 0);
+                this.data = null;
+                this.nativeData = nativeData;
+                this.nativeDataLen = nativeDataLen;
+            }
+
+            /// <summary>
+            /// Indicates whether this instance and a specified object are equal.
+            /// </summary>
+            public override bool Equals(object obj)
+            {
+                return obj is BinaryKey && Equals((BinaryKey)obj);
+            }
+
+            /// <summary>
+            /// Returns the hash code for this instance.
+            /// </summary>
+            public override int GetHashCode()
+            {
+                return ComputeHashCode(GetSpan());
+            }
+
+            /// <summary>
+            /// Indicates whether this instance and a specified object are equal.
+            /// </summary>
+            public bool Equals(BinaryKey other)
+            {
+                return GetSpan().SequenceEqual(other.GetSpan());
+            }
+
+            private ReadOnlySpan<byte> GetSpan()
```

Comment:
Possible simplification: instead of storing all of `data`, `nativeData`, and `nativeDataLen` in this `BinaryKey` object, and then generating new `ReadOnlySpan` objects from them, depending on which ones were provided in the constructor, what do you think about removing those fields and instead storing `ReadOnlySpan` objects in `BinaryKey`'s. We could then remove `GetSpan` and these `nullptr` conditionals and instead construct the right `ReadOnlySpan` objects just depending on which constructor is being called.